patroni/config_generator.py

```diff
@@ -0,0 +1,290 @@
+            with open(default_hba_path, 'r') as f:
+                config['postgresql']['pg_hba'] = list(filter(lambda i: i and i.split()[0] in allowed_records,
+                                                             (li.strip() for li in f.readlines())))
+        except OSError as e:
+            sys.exit(f'Failed to read pg_hba.conf: {e}')
+
+    default_ident_path = os.path.join(config['postgresql']['data_dir'], 'pg_ident.conf')
+    if config['postgresql']['parameters']['ident_file'] == default_ident_path:
+        try:
+            with open(default_ident_path, 'r') as f:
+                config['postgresql']['pg_ident'] = [i.strip() for i in f.readlines()
+                                                    if i.strip() and not i.startswith('#')]
+        except OSError as e:
+            sys.exit(f'Failed to read pg_ident.conf: {e}')
+        if not config['postgresql']['pg_ident']:
+            del config['postgresql']['pg_ident']
+
+    config['postgresql']['authentication'] = {
+        'superuser': su_params,
+        'replication': {'username': no_value_msg, 'password': no_value_msg}
+    }
+
+
+def generate_config(file: str, sample: bool, dsn: Optional[str]) -> None:
+    """Generate Patroni configuration file.
+
+    Gather all the available non-internal GUC values having configuration file, postmaster command line or environment
+    variable as a source and store them in the appropriate part of Patroni configuration (``postgresql.parameters`` or
+    ``bootsrtap.dcs.postgresql.parameters``). Either the provided DSN (takes precedence) or PG ENV vars will be used
+    for the connection. If password is not provided, it should be entered via prompt.
+
+    The created configuration contains:
+    - ``scope``: cluster_name GUC value or PATRONI_SCOPE ENV variable value if available
+    - ``name``: PATRONI_NAME ENV variable value if set, otherewise hostname
+    - ``bootsrtap.dcs``: section with all the parameters (incl. the majority of PG GUCs) set to their default values
+      defined by Patroni and adjusted by the source instances's configuration values.
+    - ``postgresql.parameters``: the source instance's archive_command, restore_command, archive_cleanup_command,
+      recovery_end_command, ssl_passphrase_command, hba_file, ident_file, config_file GUC values
+    - ``postgresql.bin_dir``: path to Postgres binaries gathered from the running instance or, if not available,
+      the value of PATRONI_POSTGRESQL_BIN_DIR ENV variable. Otherwise, an empty string.
+    - ``postgresql.datadir``: the value gathered from the corresponding PG GUC
+    - ``postgresql.listen``: source instance's listen_addresses and port GUC values
+    - ``postgresql.connect_address``: if possible, generated from the connection params
+    - ``postgresql.authentication``:
+        - superuser and replication users defined (if possible, usernames are set from the respective Patroni ENV vars,
```

Comment:
```suggestion
    The created configuration contains:

    - ``scope``: ``cluster_name`` GUC value or ``PATRONI_SCOPE`` ENV variable value if available
    - ``name``: ``PATRONI_NAME`` ENV variable value if set, otherwise hostname
    - ``bootsrtap.dcs``: section with all the parameters (incl. the majority of PG GUCs) set to their default values
      defined by Patroni and adjusted by the source instances's configuration values.
    - ``postgresql.parameters``: the source instance's ``archive_command``, ``restore_command``, ``archive_cleanup_command``,
      ``recovery_end_command``, ``ssl_passphrase_command``, ``hba_file``, ``ident_file``, ``config_file`` GUC values
    - ``postgresql.bin_dir``: path to Postgres binaries gathered from the running instance or, if not available,
      the value of ``PATRONI_POSTGRESQL_BIN_DIR`` ENV variable. Otherwise, an empty string.
    - ``postgresql.datadir``: the value gathered from the corresponding PG GUC
    - ``postgresql.listen``: source instance's ``listen_addresses`` and ``port`` GUC values
    - ``postgresql.connect_address``: if possible, generated from the connection params
    - ``postgresql.authentication``:

        - superuser and replication users defined (if possible, usernames are set from the respective Patroni ENV vars,
```